In [29]:
import pandas as pd

In [30]:
df = pd.read_excel("bangalore_cars.xlsx")

In [37]:
df['new_car_detail'].to_list()

["{'it': 0, 'ft': 'Petrol', 'bt': 'Hatchback', 'km': '1,20,000', 'transmission': 'Manual', 'ownerNo': 3, 'owner': '3rd Owner', 'oem': 'Maruti', 'model': 'Maruti Celerio', 'modelYear': 2015, 'centralVariantId': 3979, 'variantName': 'VXI', 'price': '₹ 4 Lakh', 'priceActual': '', 'priceSaving': '', 'priceFixedText': None, 'trendingText': {'imgUrl': 'https://stimg.cardekho.com/used-cars/common/icons/trending.svg', 'heading': 'Trending Car!', 'desc': 'High chances of sale in next 6 days'}}",
 "{'it': 0, 'ft': 'Petrol', 'bt': 'SUV', 'km': '32,706', 'transmission': 'Manual', 'ownerNo': 2, 'owner': '2nd Owner', 'oem': 'Ford', 'model': 'Ford Ecosport', 'modelYear': 2018, 'centralVariantId': 6087, 'variantName': '1.5 Petrol Titanium BSIV', 'price': '₹ 8.11 Lakh', 'priceActual': '', 'priceSaving': '', 'priceFixedText': None, 'trendingText': {'imgUrl': 'https://stimg.cardekho.com/used-cars/common/icons/trending.svg', 'heading': 'Trending Car!', 'desc': 'High chances of sale in next 6 days'}}",
 "{

In [41]:
import pandas as pd
import ast

# Sample data: assuming df is your DataFrame with a column 'new_car_detail'
data_list = df['new_car_detail'].to_list()

# Function to convert string to dictionary if necessary
def to_dict(data):
    if isinstance(data, str):
        try:
            return ast.literal_eval(data)
        except (ValueError, SyntaxError):
            return {}
    return data

# Function to flatten a single dictionary
def flatten_dict(data):
    if not isinstance(data, dict):
        return {}
    
    if 'trendingText' in data:
        # Flatten the nested dictionary
        flat_data = {
            **data,  # Unpack the main dictionary
            **{f'trendingText_{k}': v for k, v in data['trendingText'].items()}  # Flatten the nested dictionary
        }
        del flat_data['trendingText']  # Remove the original nested dictionary
    else:
        # If 'trendingText' key is missing, just return the data as is
        flat_data = data
    return flat_data

# Convert strings to dictionaries if needed
data_list = [to_dict(item) for item in data_list]

# Flatten all rows
flattened_data_list = [flatten_dict(row) for row in data_list]

# Convert to DataFrame
df_flattened = pd.DataFrame(flattened_data_list)

# Display the DataFrame
df_flattened


it      ft         bt        km transmission  ownerNo      owner  \
0      0  Petrol  Hatchback  1,20,000       Manual        3  3rd Owner   
1      0  Petrol        SUV    32,706       Manual        2  2nd Owner   
2      0  Petrol  Hatchback    11,949       Manual        1  1st Owner   
3      0  Petrol      Sedan    17,794       Manual        1  1st Owner   
4      0  Diesel        SUV    60,000       Manual        1  1st Owner   
...   ..     ...        ...       ...          ...      ...        ...   
1476   0  Diesel        SUV  1,56,039       Manual        2  2nd Owner   
1477   0  Petrol      Sedan    56,000       Manual        2  2nd Owner   
1478   0  Petrol  Hatchback    42,000       Manual        2  2nd Owner   
1479   0  Diesel  Hatchback    93,003       Manual        1  1st Owner   
1480   0  Petrol  Hatchback    36,000    Automatic        2  2nd Owner   

          oem               model  modelYear  centralVariantId  \
0      Maruti      Maruti Celerio       2015              3979   
1        Ford       Ford Ecosport       2018              6087   
2        Tata          Tata Tiago       2018              2983   
3     Hyundai       Hyundai Xcent       2014              1867   
4      Maruti  Maruti SX4 S Cross       2015              4277   
...       ...                 ...        ...               ...   
1476   Toyota     Toyota Fortuner       2012              2253   
1477   Maruti          Maruti SX4       2008              4257   
1478    Honda          Honda Brio       2014              1188   
1479  Hyundai         Hyundai i20       2018              5695   
1480  Hyundai   Hyundai Grand i10       2017              1523   

                   variantName         price priceActual priceSaving  \
0                          VXI      ₹ 4 Lakh                           
1     1.5 Petrol Titanium BSIV   ₹ 8.11 Lakh                           
2              1.2 Revotron XZ   ₹ 5.85 Lakh                           
3           1.2 Kappa S Option   ₹ 4.62 Lakh                           
4                DDiS 200 Zeta   ₹ 7.90 Lakh                           
...                        ...           ...         ...         ...   
1476                    4x4 MT  ₹ 16.49 Lakh                           
1477                 Zxi BSIII   ₹ 3.30 Lakh                           
1478                      S MT   ₹ 4.25 Lakh                           
1479        1.4 Asta Dual Tone   ₹ 7.50 Lakh                           
1480            Asta Option AT   ₹ 6.75 Lakh                           

     priceFixedText                                trendingText_imgUrl  \
0              None  https://stimg.cardekho.com/used-cars/common/ic...   
1              None  https://stimg.cardekho.com/used-cars/common/ic...   
2              None  https://stimg.cardekho.com/used-cars/common/ic...   
3              None  https://stimg.cardekho.com/used-cars/common/ic...   
4              None  https://stimg.cardekho.com/used-cars/common/ic...   
...             ...                                                ...   
1476           None  https://stimg.cardekho.com/used-cars/common/ic...   
1477           None  https://stimg.cardekho.com/used-cars/common/ic...   
1478           None  https://stimg.cardekho.com/used-cars/common/ic...   
1479           None  https://stimg.cardekho.com/used-cars/common/ic...   
1480           None  https://stimg.cardekho.com/used-cars/common/ic...   

     trendingText_heading                    trendingText_desc  
0           Trending Car!  High chances of sale in next 6 days  
1           Trending Car!  High chances of sale in next 6 days  
2           Trending Car!  High chances of sale in next 6 days  
3           Trending Car!  High chances of sale in next 6 days  
4           Trending Car!  High chances of sale in next 6 days  
...                   ...                                  ...  
1476        Trending Car!  High chances of sale in next 6 days  
1477        Trending Car!  High chances of sale in

In [78]:
import pandas as pd
import ast

data = df['new_car_overview'].to_list()
nested_dict_list_str= pd.DataFrame(data)

# Convert each string to a dictionary
nested_dict_list = [ast.literal_eval(d) for d in nested_dict_list_str[0]]

# Function to extract and flatten each dictionary
def extract_flatten_dict(nested_dict):
    if 'top' in nested_dict:
        top_data = nested_dict['top']
        df = pd.DataFrame(top_data)
        df = df.drop(columns=['icon'], errors='ignore')  # Drop the 'icon' column if present
        return df.set_index('key').T
    else:
        return pd.DataFrame()

# Extract and flatten all dictionaries
df_list = [extract_flatten_dict(d) for d in nested_dict_list]

# Concatenate all DataFrames into one
df_combined = pd.concat(df_list, ignore_index=True)

# Display the combined DataFrame
df_combined


key  Registration Year     Insurance Validity Fuel Type    Seats  \
0                 2015  Third Party insurance    Petrol  5 Seats   
1             Feb 2018          Comprehensive    Petrol  5 Seats   
2            Sept 2018          Comprehensive    Petrol  5 Seats   
3             Dec 2014          Comprehensive    Petrol  5 Seats   
4                 2015  Third Party insurance    Diesel  5 Seats   
...                ...                    ...       ...      ...   
1476          May 2012          Comprehensive    Diesel  7 Seats   
1477          Jun 2008  Third Party insurance    Petrol  5 Seats   
1478         Sept 2014          Comprehensive    Petrol  5 Seats   
1479          Dec 2018          Comprehensive    Diesel  5 Seats   
1480          May 2017          Comprehensive    Petrol  5 Seats   

key     Kms Driven   RTO     Ownership Engine Displacement Transmission  \
0     1,20,000 Kms  KA51   Third Owner              998 cc       Manual   
1       32,706 Kms  KA05  Second Owner             1497 cc       Manual   
2       11,949 Kms  KA03   First Owner             1199 cc       Manual   
3       17,794 Kms  KA53   First Owner             1197 cc       Manual   
4       60,000 Kms  KA04   First Owner             1248 cc       Manual   
...            ...   ...           ...                 ...          ...   
1476  1,56,039 Kms  KA01  Second Owner             2982 cc       Manual   
1477    56,000 Kms  KA02  Second Owner             1586 cc       Manual   
1478    42,000 Kms  KA03  Second Owner             1198 cc       Manual   
1479    93,003 Kms  KA03   First Owner             1396 cc       Manual   
1480    36,000 Kms  KA04  Second Owner             1197 cc    Automatic   

key  Year of Manufacture  
0                   2015  
1                   2018  
2                   2018  
3                   2014  
4                   2015  
...                  ...  
1476                2012  
1477                2008  
1478                2014  
1479                2018  
1480                2017  

[1481 rows x 10 columns]

In [80]:
df_list

[key   Registration Year     Insurance Validity Fuel Type    Seats  \
 value              2015  Third Party insurance    Petrol  5 Seats   
 
 key      Kms Driven   RTO    Ownership Engine Displacement Transmission  \
 value  1,20,000 Kms  KA51  Third Owner              998 cc       Manual   
 
 key   Year of Manufacture  
 value                2015  ,
 key   Registration Year Insurance Validity Fuel Type    Seats  Kms Driven  \
 value          Feb 2018      Comprehensive    Petrol  5 Seats  32,706 Kms   
 
 key     RTO     Ownership Engine Displacement Transmission Year of Manufacture  
 value  KA05  Second Owner             1497 cc       Manual                2018  ,
 key   Registration Year Insurance Validity Fuel Type    Seats  Kms Driven  \
 value         Sept 2018      Comprehensive    Petrol  5 Seats  11,949 Kms   
 
 key     RTO    Ownership Engine Displacement Transmission Year of Manufacture  
 value  KA03  First Owner             1199 cc       Manual                2018  ,
 

In [66]:
nested_dict_list[:10]

[{'heading': 'Car overview',
  'top': [{'key': 'Registration Year',
    'value': '2015',
    'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/registrationYear.svg'},
   {'key': 'Insurance Validity',
    'value': 'Third Party insurance',
    'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/insuranceValidity.svg'},
   {'key': 'Fuel Type',
    'value': 'Petrol',
    'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/fuel.svg'},
   {'key': 'Seats',
    'value': '5 Seats',
    'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/seats.svg'},
   {'key': 'Kms Driven',
    'value': '1,20,000 Kms',
    'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/kmsDriven.svg'},
   {'key': 'RTO',
    'value': 'KA51',
    'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/rto.svg'},
   {'key': 'Ownership',
    'value': 'Third Owner',
    'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/ownership.svg'},
   {'key': 'Engine Displacement',
    'value': '998 cc',
    'icon': 'https://i

In [9]:
df.columns

Index(['new_car_detail', 'new_car_overview', 'new_car_feature',
       'new_car_specs', 'car_links'],
      dtype='object')

In [12]:
df['new_car_detail'][0]

"{'it': 0, 'ft': 'Petrol', 'bt': 'Hatchback', 'km': '1,20,000', 'transmission': 'Manual', 'ownerNo': 3, 'owner': '3rd Owner', 'oem': 'Maruti', 'model': 'Maruti Celerio', 'modelYear': 2015, 'centralVariantId': 3979, 'variantName': 'VXI', 'price': '₹ 4 Lakh', 'priceActual': '', 'priceSaving': '', 'priceFixedText': None, 'trendingText': {'imgUrl': 'https://stimg.cardekho.com/used-cars/common/icons/trending.svg', 'heading': 'Trending Car!', 'desc': 'High chances of sale in next 6 days'}}"

In [10]:
df['new_car_overview'][0]

"{'heading': 'Car overview', 'top': [{'key': 'Registration Year', 'value': '2015', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/registrationYear.svg'}, {'key': 'Insurance Validity', 'value': 'Third Party insurance', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/insuranceValidity.svg'}, {'key': 'Fuel Type', 'value': 'Petrol', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/fuel.svg'}, {'key': 'Seats', 'value': '5 Seats', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/seats.svg'}, {'key': 'Kms Driven', 'value': '1,20,000 Kms', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/kmsDriven.svg'}, {'key': 'RTO', 'value': 'KA51', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/rto.svg'}, {'key': 'Ownership', 'value': 'Third Owner', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/ownership.svg'}, {'key': 'Engine Displacement', 'value': '998 cc', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/engineDisplacement.svg'}, {'key': 'Transmission', '

In [16]:
df['new_car_feature'][0]

"{'heading': 'Features', 'top': [{'value': 'Power Steering'}, {'value': 'Power Windows Front'}, {'value': 'Air Conditioner'}, {'value': 'Heater'}, {'value': 'Adjustable Head Lights'}, {'value': 'Manually Adjustable Exterior Rear View Mirror'}, {'value': 'Centeral Locking'}, {'value': 'Child Safety Locks'}], 'data': [{'heading': 'Comfort & Convenience', 'subHeading': 'Comfort', 'list': [{'value': 'Power Steering'}, {'value': 'Power Windows Front'}, {'value': 'Power Windows Rear'}, {'value': 'Remote Trunk Opener'}, {'value': 'Remote Fuel Lid Opener'}, {'value': 'Low Fuel Warning Light'}, {'value': 'Accessory Power Outlet'}, {'value': 'Vanity Mirror'}, {'value': 'Rear Seat Headrest'}, {'value': 'Cup Holders Front'}]}, {'heading': 'Interior', 'subHeading': 'Interior', 'list': [{'value': 'Air Conditioner'}, {'value': 'Heater'}, {'value': 'Digital Odometer'}, {'value': 'Electronic Multi Tripmeter'}, {'value': 'Fabric Upholstery'}, {'value': 'Glove Compartment'}, {'value': 'Digital Clock'}]},

In [17]:
df['new_car_specs'][0]

"{'heading': 'Specifications', 'top': [{'key': 'Mileage', 'value': '23.1 kmpl'}, {'key': 'Engine', 'value': '998 CC'}, {'key': 'Max Power', 'value': '67.04bhp'}, {'key': 'Torque', 'value': '90Nm'}, {'key': 'Seats', 'value': '5'}], 'data': [{'heading': 'Engine and Transmission', 'subHeading': 'Engine', 'list': [{'key': 'Color', 'value': 'White'}, {'key': 'Engine Type', 'value': 'K10B Engine'}, {'key': 'Displacement', 'value': '998'}, {'key': 'Max Power', 'value': '67.04bhp@6000rpm'}, {'key': 'Max Torque', 'value': '90Nm@3500rpm'}, {'key': 'No of Cylinder', 'value': 3}, {'key': 'Values per Cylinder', 'value': 4}, {'key': 'Value Configuration', 'value': 'DOHC'}, {'key': 'Fuel Suppy System', 'value': 'MPFi'}, {'key': 'BoreX Stroke', 'value': '73 X 82 mm'}, {'key': 'Compression Ratio', 'value': '11.0:1'}, {'key': 'Turbo Charger', 'value': 'No'}, {'key': 'Super Charger', 'value': 'No'}]}, {'heading': 'Dimensions & Capacity', 'subHeading': 'Dimensions', 'list': [{'key': 'Length', 'value': '37

In [18]:
df['car_links'][0]

'https://www.cardekho.com/used-car-details/used-Maruti-Celerio-Vxi-cars-Bangalore_a12b21b4-e26a-41bf-8f47-7e79a725a246.htm'

df.columns = ['new_car_detail', 'new_car_overview', 'new_car_feature',
       'new_car_specs', 'car_links']

df['new_car_detail'][0] =  "{'it': 0, 'ft': 'Petrol', 'bt': 'Hatchback', 'km': '1,20,000', 'transmission': 'Manual', 'ownerNo': 3, 'owner': '3rd Owner', 'oem': 'Maruti', 'model': 'Maruti Celerio', 'modelYear': 2015, 'centralVariantId': 3979, 'variantName': 'VXI', 'price': '₹ 4 Lakh', 'priceActual': '', 'priceSaving': '', 'priceFixedText': None, 'trendingText': {'imgUrl': 'https://stimg.cardekho.com/used-cars/common/icons/trending.svg', 'heading': 'Trending Car!', 'desc': 'High chances of sale in next 6 days'}}"

df['new_car_overview'][0] = "{'heading': 'Car overview', 'top': [{'key': 'Registration Year', 'value': '2015', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/registrationYear.svg'}, {'key': 'Insurance Validity', 'value': 'Third Party insurance', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/insuranceValidity.svg'}, {'key': 'Fuel Type', 'value': 'Petrol', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/fuel.svg'}, {'key': 'Seats', 'value': '5 Seats', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/seats.svg'}, {'key': 'Kms Driven', 'value': '1,20,000 Kms', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/kmsDriven.svg'}, {'key': 'RTO', 'value': 'KA51', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/rto.svg'}, {'key': 'Ownership', 'value': 'Third Owner', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/ownership.svg'}, {'key': 'Engine Displacement', 'value': '998 cc', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/engineDisplacement.svg'}, {'key': 'Transmission', 'value': 'Manual', 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/transmission.svg'}, {'key': 'Year of Manufacture', 'value': 2015, 'icon': 'https://images10.gaadi.com/listing/vdp/co/v1/yearManufacture.svg'}], 'bottomData': None}"

df['new_car_feature'][0] = "{'heading': 'Features', 'top': [{'value': 'Power Steering'}, {'value': 'Power Windows Front'}, {'value': 'Air Conditioner'}, {'value': 'Heater'}, {'value': 'Adjustable Head Lights'}, {'value': 'Manually Adjustable Exterior Rear View Mirror'}, {'value': 'Centeral Locking'}, {'value': 'Child Safety Locks'}], 'data': [{'heading': 'Comfort & Convenience', 'subHeading': 'Comfort', 'list': [{'value': 'Power Steering'}, {'value': 'Power Windows Front'}, {'value': 'Power Windows Rear'}, {'value': 'Remote Trunk Opener'}, {'value': 'Remote Fuel Lid Opener'}, {'value': 'Low Fuel Warning Light'}, {'value': 'Accessory Power Outlet'}, {'value': 'Vanity Mirror'}, {'value': 'Rear Seat Headrest'}, {'value': 'Cup Holders Front'}]}, {'heading': 'Interior', 'subHeading': 'Interior', 'list': [{'value': 'Air Conditioner'}, {'value': 'Heater'}, {'value': 'Digital Odometer'}, {'value': 'Electronic Multi Tripmeter'}, {'value': 'Fabric Upholstery'}, {'value': 'Glove Compartment'}, {'value': 'Digital Clock'}]}, {'heading': 'Exterior', 'subHeading': 'Exterior', 'list': [{'value': 'Adjustable Head Lights'}, {'value': 'Manually Adjustable Exterior Rear View Mirror'}, {'value': 'Wheel Covers'}, {'value': 'Power Antenna'}, {'value': 'Chrome Grille'}]}, {'heading': 'Safety', 'subHeading': 'Safety', 'list': [{'value': 'Centeral Locking'}, {'value': 'Child Safety Locks'}, {'value': 'Day Night Rear View Mirror'}, {'value': 'Passenger Side Rear View Mirror'}, {'value': 'Halogen Headlamps'}, {'value': 'Rear Seat Belts'}, {'value': 'Door Ajar Warning'}, {'value': 'Side Impact Beams'}, {'value': 'Front Impact Beams'}, {'value': 'Adjustable Seats'}, {'value': 'Centrally Mounted Fuel Tank'}, {'value': 'Engine Immobilizer'}, {'value': 'Anti Theft Device'}]}], 'commonIcon': 'https://stimg.cardekho.com/pwa/img/vdpN/tickG.svg'}"

df['new_car_specs'][0] = "{'heading': 'Specifications', 'top': [{'key': 'Mileage', 'value': '23.1 kmpl'}, {'key': 'Engine', 'value': '998 CC'}, {'key': 'Max Power', 'value': '67.04bhp'}, {'key': 'Torque', 'value': '90Nm'}, {'key': 'Seats', 'value': '5'}], 'data': [{'heading': 'Engine and Transmission', 'subHeading': 'Engine', 'list': [{'key': 'Color', 'value': 'White'}, {'key': 'Engine Type', 'value': 'K10B Engine'}, {'key': 'Displacement', 'value': '998'}, {'key': 'Max Power', 'value': '67.04bhp@6000rpm'}, {'key': 'Max Torque', 'value': '90Nm@3500rpm'}, {'key': 'No of Cylinder', 'value': 3}, {'key': 'Values per Cylinder', 'value': 4}, {'key': 'Value Configuration', 'value': 'DOHC'}, {'key': 'Fuel Suppy System', 'value': 'MPFi'}, {'key': 'BoreX Stroke', 'value': '73 X 82 mm'}, {'key': 'Compression Ratio', 'value': '11.0:1'}, {'key': 'Turbo Charger', 'value': 'No'}, {'key': 'Super Charger', 'value': 'No'}]}, {'heading': 'Dimensions & Capacity', 'subHeading': 'Dimensions', 'list': [{'key': 'Length', 'value': '3715mm'}, {'key': 'Width', 'value': '1635mm'}, {'key': 'Height', 'value': '1565mm'}, {'key': 'Wheel Base', 'value': '2425mm'}, {'key': 'Front Tread', 'value': '1420mm'}, {'key': 'Rear Tread', 'value': '1410mm'}, {'key': 'Kerb Weight', 'value': '835kg'}, {'key': 'Gross Weight', 'value': '1250kg'}]}, {'heading': 'Miscellaneous', 'subHeading': 'Miscellaneous', 'list': [{'key': 'Gear Box', 'value': '5 Speed '}, {'key': 'Drive Type', 'value': 'FWD'}, {'key': 'Seating Capacity', 'value': '5'}, {'key': 'Steering Type', 'value': 'Power'}, {'key': 'Turning Radius', 'value': '4.7 metres'}, {'key': 'Front Brake Type', 'value': 'Ventilated Disc'}, {'key': 'Rear Brake Type', 'value': 'Drum'}, {'key': 'Top Speed', 'value': '150 Kmph'}, {'key': 'Acceleration', 'value': '15.05 Seconds'}, {'key': 'Tyre Type', 'value': 'Tubeless, Radial'}, {'key': 'No Door Numbers', 'value': '5'}, {'key': 'Cargo Volumn', 'value': '235-litres'}]}], 'commonIcon': ''}"

In [27]:
df.columns

Index(['new_car_detail', 'new_car_overview', 'new_car_feature',
       'new_car_specs', 'car_links'],
      dtype='object')